ic-agent/src/hash_tree.rs

```diff
@@ -0,0 +1,274 @@
+/** Hash-tree
+
+#### Resources:
+
+- [Public spec link](https://hydra.dfinity.systems/latest/dfinity-ci-build/ic-ref.pr-218/public-spec/1/index.html#_encoding_of_certificates)
+
+- [Reference implementation in the replica](https://github.com/dfinity-lab/dfinity/tree/master/rs/crypto/tree_hash)
+
+*/
+use serde::{export::Formatter, Deserialize, Serialize};
+
+use openssl::sha::Sha256;
+
+use std::convert::TryFrom;
+use std::convert::TryInto;
+use std::fmt;
+use std::fmt::Debug;
+
+const DOMAIN_HASHTREE_LEAF: &str = "ic-hashtree-leaf";
+const DOMAIN_HASHTREE_EMPTY_SUBTREE: &str = "ic-hashtree-empty";
+const DOMAIN_HASHTREE_NODE: &str = "ic-hashtree-labeled";
+const DOMAIN_HASHTREE_FORK: &str = "ic-hashtree-fork";
+
+/// A blob used as a label in the tree.
+///
+/// Most labels are expected to be printable ASCII strings, but some
+/// are just short sequences of arbitrary bytes (e.g., CanisterIds).
+#[derive(Clone, Serialize, Deserialize)]
+#[serde(from = "&serde_bytes::Bytes")]
+#[serde(into = "serde_bytes::ByteBuf")]
+pub struct Label(LabelRepr);
+
+/// Represents a path in a hash tree.
+pub type Path = Vec<Label>;
+
+/// Vec<u8> is typically 3 machine words long (pointer + size + capacity) which
+/// is 24 bytes on amd64.  It's a good practice to keep enum variants of
+/// approximately the same size. We want to optimize for labels of at most 32
+/// bytes (as we will have many labels that are SHA256 hashes).
+const SMALL_LABEL_SIZE: usize = 32;
+
+/// This type hides the implementation of Label.
+#[derive(Clone)]
+enum LabelRepr {
```

Comment:
Why do we need this type? If you want to save RAM, we could just use https://docs.rs/smallvec/1.4.2/smallvec/ ?

Reply:
No real gain to your suggestion, just premature optimization.

Inductive structures are easier to understand, and more closely follow the public spec.

The code here adapts somewhat more complex replica code that does the same thing.  Argue with them if you really care.  I don't think we can afford to.